bazel/generate_objc.bzl

```diff
@@ -0,0 +1,225 @@
+            outs += [_get_output_file_name_from_proto(proto, _GRPC_PROTO_SRC_FMT)]
+    
+    out_files = [ctx.actions.declare_file(out) for out in outs]
+    dir_out = _join_directories([
+        str(ctx.genfiles_dir.path), label_package, _GENERATED_PROTOS_DIR
+    ])
+
+    arguments = []
+    if ctx.executable.plugin:
+        arguments += get_plugin_args(
+            ctx.executable.plugin,
+            [],
+            dir_out,
+            False,
+        )
+        tools = [ctx.executable.plugin]
+    arguments += ["--objc_out=" + dir_out]
+
+    arguments += ["--proto_path=."]
+    arguments += get_include_protoc_args(protos)
+    # Include the output directory so that protoc puts the generated code in the
+    # right directory.
+    arguments += ["--proto_path={}".format(dir_out)]
+    arguments += ["--proto_path={}".format(_get_directory_from_proto(proto)) for proto in protos]
+    arguments += [_get_srcs_file_path(proto) for proto in protos]
+
+    # create a list of well known proto files if the argument is non-None
+    well_known_proto_files = []
+    if ctx.attr.use_well_known_protos:
+        f = ctx.attr.well_known_protos.files.to_list()[0].dirname
+        arguments += ["-I{0}".format(f + "/../..")]
+        well_known_proto_files = ctx.attr.well_known_protos.files.to_list()
+    ctx.actions.run(
+        inputs = protos + well_known_proto_files,
+        tools = tools,
+        outputs = out_files,
+        executable = ctx.executable._protoc,
+        arguments = arguments,
+    )
+
+    return struct(files = depset(out_files))
+
+def _get_output_file_name_from_proto(proto, fmt):
+    return proto_path_to_generated_filename(
+        _GENERATED_PROTOS_DIR + "/" +
+        _get_directory_from_proto(proto) + _get_slash_or_null_from_proto(proto) +
+        to_upper_camel_with_extension(_get_file_name_from_proto(proto), "proto"),
+        fmt,
+    )
+
+def _get_file_name_from_proto(proto):
+    return proto.path.rpartition("/")[2]
+
+def _get_slash_or_null_from_proto(proto):
+    """Potentially returns empty (if the file is in the root directory)"""
+    return proto.path.rpartition("/")[1]
+
+def _get_directory_from_proto(proto):
+    return proto.path.rpartition("/")[0]
+
+def _strip_package_and_to_camel_case(label_package, file):
+    return to_upper_camel_with_extension(
+        _strip_package_from_path(label_package, file),
+        "proto"
+    )
+
+def _strip_package_from_path(label_package, file):
+    """strips the label's package from the file path
+    e.g. In package src/objective-c: src/objective-c/foo/bar.proto -> foo/bar.proto
+    If the file is not in the same package, then the files path prepended with "//" is returned
+    """
+    prefix_len = 0
+    if not file.is_source and file.path.startswith(file.root.path):
+        prefix_len = len(file.root.path) + 1
+
+    path = file.path
+    file_name = ""
+    if len(label_package) == 0:
+        file_name = path
```

Comment:
`return path`

Reply:
`file.path` could also include the prefix? is that ok?

Reply:
`src/objective-c/foo.proto` -> `foo.proto` but `examples/protos/bar.proto` should be `//examples/protos/bar.proto`, because we want absolute paths for proto files from different packages. Does it make sense to you?